[Bug] [Fix and code suggested] EM_ASM does not allow modifying operations as arguments

# The issue
Consider the following snippet compiled with `em++ a.cpp` (`
emcc (Emscripten gcc/clang-like replacement) 1.38.38 (commit 49652605051d9faca2dcb781918a028bdfbc2703)`):

```c++
#include <emscripten/emscripten.h>

int main() {
    int x = 0, y = 0;
    EM_ASM({ console.log($0, $1); }, x++, y++);
}
```

**Expected output**: `0 0`
**Real output**: `1 1`

## The reason
The following warning is produced:

```
a.cpp:5:39: warning: multiple unsequenced modifications to 'x' [-Wunsequenced]
    EM_ASM({ console.log($0, $1); }, x++, y++);
                                      ^~
c:\emsdk\upstream\emscripten\system\include\emscripten/em_asm.h:176:83: note: expanded from macro 'EM_ASM'
#define EM_ASM(code, ...) ((void)emscripten_asm_const_int(#code _EM_ASM_PREP_ARGS(__VA_ARGS__)))
                                                                                  ^~~~~~~~~~~
c:\emsdk\upstream\emscripten\system\include\emscripten/em_asm.h:122:42: note: expanded from macro '_EM_ASM_PREP_ARGS'
    , __em_asm_sig_builder::__em_asm_sig(__VA_ARGS__).buffer, ##__VA_ARGS__
                                         ^                      ~~~~~~~~~~~
a.cpp:5:44: warning: multiple unsequenced modifications to 'y' [-Wunsequenced]
    EM_ASM({ console.log($0, $1); }, x++, y++);
                                           ^~
c:\emsdk\upstream\emscripten\system\include\emscripten/em_asm.h:176:83: note: expanded from macro 'EM_ASM'
#define EM_ASM(code, ...) ((void)emscripten_asm_const_int(#code _EM_ASM_PREP_ARGS(__VA_ARGS__)))
                                                                                  ^~~~~~~~~~~
c:\emsdk\upstream\emscripten\system\include\emscripten/em_asm.h:122:42: note: expanded from macro '_EM_ASM_PREP_ARGS'
    , __em_asm_sig_builder::__em_asm_sig(__VA_ARGS__).buffer, ##__VA_ARGS__
                                         ^                      ~~~~~~~~~~~
2 warnings generated.
```

Specifically, when `EM_ASM` is expanded in C++ mode with non-asmjs backend, it's expanded to:

```c++
((void)emscripten_asm_const_int("{ console.log($0, $1); }" , __em_asm_sig_builder::__em_asm_sig(x++, y++).buffer,x++, y++));
```

Here we clearly see the unsequenced modification which is undefined behaviour in C++: each argument is evaluted twice: once to pass to `__em_asm_sig` to examine the type of the argument and build corresponding type signature, and once to `emscripten_asm_const_int` to pass the actual value.

## Severity

On the one hand, it's a warning enabled by default, so users may investigate further and avoid undesired consequences. Fortunately, it looks like more complex types (like vectors) are not supported by `EM_ASM` at all, so the warning _probably_ catches all cases.

On the other hand, this bug occured in [`asm-dom`](https://github.com/mbasso/asm-dom/blob/ea3dba7b0e031051eb2b8aa2eba53d99912305ab/src/cpp/Patch/patch.cpp#L91-L93) at the very least, even though it's compiled with `-Wall -Werror`. So it looks like this warning evaded the developer in that specific case.

# Suggestion: quick fix
Document that `EM_ASM`'s arguments are not to have any side effects at all.

# Suggestion: full fix

Values passed to [`__em_asm_sig`](https://github.com/emscripten-core/emscripten/blob/941bbc6b9b35d3124f17d2503d7a32cc81032dac/system/include/emscripten/em_asm.h#L112-L119) are not really needed: we only want to know their types. We can do that by passing arguments to `decltype`, which prevents evaluation of these arguments, thus removing unsequenced modifications. As we have to pass several arguments, we can wrap them into `std::make_tuple` (not exactly sure how that will play with constness and stuff, but probably in a reasonable way because only trivial types are passed, essentially).

Some rewriting of `__em_asm_sig_builder` should happen as well to work strictly on type-level so it can be reused in the future for values which cannot be instantiated simply (note that `std::declval()`'s result cannot be used in any evaluated expression even if it's discarded).

Here is my attempt which monkey-patches emscripten, you can compile it with `em++ a.cpp -std=c+=11` and see the expected `0 0`:

```c++
#include <emscripten.h>
#include <tuple> // std::tuple, std::make_tuple
#include <type_traits> // std:integral_constant

// Function overload requires passing a value of a type, but we
// may have no easy way to instantiate a value of a generic type.
template<typename> struct __em_asm_sig {};
template<> struct __em_asm_sig<float> : std::integral_constant<char, 'd'> {};
template<> struct __em_asm_sig<double> : std::integral_constant<char, 'd'> {};
template<> struct __em_asm_sig<int> : std::integral_constant<char, 'i'> {};
template<> struct __em_asm_sig<unsigned> : std::integral_constant<char, 'i'> {};
template<> struct __em_asm_sig<long> : std::integral_constant<char, 'i'> {};
template<> struct __em_asm_sig<unsigned long> : std::integral_constant<char, 'i'> {};
template<typename T> struct __em_asm_sig<T*> : std::integral_constant<char, 'i'> {};

template<typename>
struct __em_asm_sig_builder2 {};

template<typename... Args>
struct __em_asm_sig_builder2<std::tuple<Args...>> {
  // No need for internal buffer or copying, just initialize directly.
  char buffer[sizeof...(Args)+1] = { __em_asm_sig<Args>::value..., 0 };
};

// Redefine old macro to use the new builder instead
#undef _EM_ASM_PREP_ARGS
#define _EM_ASM_PREP_ARGS(...) \
    , __em_asm_sig_builder2<decltype(std::make_tuple(__VA_ARGS__))>().buffer, ##__VA_ARGS__

int main() {
    int x = 0, y = 0;
    EM_ASM({ console.log($0, $1); }, x++, y++);
}
```

I hope that most of the example above can be a drop-in replacement of [`__em_asm_sig_builder`](https://github.com/emscripten-core/emscripten/blob/941bbc6b9b35d3124f17d2503d7a32cc81032dac/system/include/emscripten/em_asm.h#L81-L123).

Here, `EM_ASM` unwraps to

```
((void)emscripten_asm_const_int("{ console.log($0, $1); }" , __em_asm_sig_builder2<decltype(std::make_tuple(x++, y++))>().buffer,x++, y++));
```

and no unordered modifications occur, as the first `x++` happens inside `decltype` which prevents the subexpression from evaluating.

--- a/system/include/emscripten/em_asm.h
+++ b/system/include/emscripten/em_asm.h
@@ -122,14 +122,17 @@ struct __em_asm_sig_builder {};
 
 template<typename... Args>
 struct __em_asm_sig_builder<__em_asm_type_tuple<Args...> > {
-  char buffer[sizeof...(Args)+1] = { __em_asm_sig<Args>::value..., 0 };
+  static const char buffer[sizeof...(Args) + 1];
 };
+
+template<typename... Args>
+const char __em_asm_sig_builder<__em_asm_type_tuple<Args...> >::buffer[] = { __em_asm_sig<Args>::value..., 0 };
 
 // We move to type level with decltype(make_tuple(...)) to avoid double
 // evaluation of arguments. Use __typeof__ instead of decltype, though,
 // because the header should be able to compile with clang's -std=c++03.
 #define _EM_ASM_PREP_ARGS(...) \
-    , __em_asm_sig_builder<__typeof__(__em_asm_make_type_tuple(__VA_ARGS__))>().buffer, ##__VA_ARGS__
+    , __em_asm_sig_builder<__typeof__(__em_asm_make_type_tuple(__VA_ARGS__))>::buffer, ##__VA_ARGS__
 
 extern "C" {
 #endif // __cplusplus

--- a/system/include/emscripten/em_asm.h
+++ b/system/include/emscripten/em_asm.h
@@ -92,52 +92,44 @@
 // C++ needs to support vararg template parameter packs, e.g. like in
 // tests/core/test_em_asm_parameter_pack.cpp. Because of that, a macro-only
 // approach doesn't work (a macro applied to a parameter pack would expand
-// incorrectly). So we can use a template function instead to build a
-// std::string, and convert that to a C string.
-// String builder class is so the _sig functions can be mutually recursive.
-class __em_asm_sig_builder {
-private:
-  static char sig_char(float) {
-    return 'd';
-  }
-  static char sig_char(double) {
-    return 'd';
-  }
-  static char sig_char(int) {
-    return 'i';
-  }
-  static char sig_char(unsigned) {
-    return 'i';
-  }
-  static char sig_char(long) {
-    return 'i';
-  }
-  static char sig_char(unsigned long) {
-    return 'i';
-  }
-  template <typename T>
-  static char sig_char(T *arg) {
-    return 'i';
-  }
+// incorrectly). So we can use a template class instead to build a temporary
+// buffer of characters.
 
-  template <typename ...Args>
-  struct inner {
-    char buffer[sizeof...(Args)+1];
-  };
-public:
-  template <typename ...Args>
-  static const inner<Args...> __em_asm_sig(Args ...args) {
-    inner<Args...> temp;
-    char buf[sizeof...(Args)+1] = { sig_char(args)..., 0 };
-    for (int i = 0; i < sizeof...(Args)+1; ++i) {
-        temp.buffer[i] = buf[i];
-    }
-    return temp;
-  }
+// As emscripten is require to build successfully with -std=c++03, we cannot
+// use std::tuple or std::integral_constant. Using C++11 features is only a
+// warning in modern Clang, which are ignored in system headers.
+template<typename> struct __em_asm_sig {};
+template<> struct __em_asm_sig<float> { static const char value = 'd'; };
+template<> struct __em_asm_sig<double> { static const char value = 'd'; };
+template<> struct __em_asm_sig<int> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned> { static const char value = 'i'; };
+template<> struct __em_asm_sig<long> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned long> { static const char value = 'i'; };
+template<typename T> struct __em_asm_sig<T*> { static const char value = 'i'; };
+
+// Instead of std::tuple
+template<typename... Args>
+struct __em_asm_type_tuple {};
+
+// Instead of std::make_tuple
+template<typename... Args>
+__em_asm_type_tuple<Args...> __em_asm_make_type_tuple(Args... args) {
+    return {};
+}
+
+template<typename>
+struct __em_asm_sig_builder {};
+
+template<typename... Args>
+struct __em_asm_sig_builder<__em_asm_type_tuple<Args...> > {
+  char buffer[sizeof...(Args)+1] = { __em_asm_sig<Args>::value..., 0 };
 };
 
+// We move to type level with decltype(make_tuple(...)) to avoid double
+// evaluation of arguments. Use __typeof__ instead of decltype, though,
+// because the header should be able to compile with clang's -std=c++03.
 #define _EM_ASM_PREP_ARGS(...) \
-    , __em_asm_sig_builder::__em_asm_sig(__VA_ARGS__).buffer, ##__VA_ARGS__
+    , __em_asm_sig_builder<__typeof__(__em_asm_make_type_tuple(__VA_ARGS__))>().buffer, ##__VA_ARGS__
 
 extern "C" {
 #endif // __cplusplus

--- a/system/include/emscripten/em_asm.h
+++ b/system/include/emscripten/em_asm.h
@@ -101,10 +101,17 @@
 template<typename> struct __em_asm_sig {};
 template<> struct __em_asm_sig<float> { static const char value = 'd'; };
 template<> struct __em_asm_sig<double> { static const char value = 'd'; };
+template<> struct __em_asm_sig<char> { static const char value = 'i'; };
+template<> struct __em_asm_sig<signed char> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned char> { static const char value = 'i'; };
+template<> struct __em_asm_sig<short> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned short> { static const char value = 'i'; };
 template<> struct __em_asm_sig<int> { static const char value = 'i'; };
-template<> struct __em_asm_sig<unsigned> { static const char value = 'i'; };
+template<> struct __em_asm_sig<unsigned int> { static const char value = 'i'; };
 template<> struct __em_asm_sig<long> { static const char value = 'i'; };
 template<> struct __em_asm_sig<unsigned long> { static const char value = 'i'; };
+template<> struct __em_asm_sig<bool> { static const char value = 'i'; };
+template<> struct __em_asm_sig<wchar_t> { static const char value = 'i'; };
 template<typename T> struct __em_asm_sig<T*> { static const char value = 'i'; };
 
 // Instead of std::tuple

--- a/system/include/emscripten/em_asm.h
+++ b/system/include/emscripten/em_asm.h
@@ -98,7 +98,7 @@
 // As emscripten is require to build successfully with -std=c++03, we cannot
 // use std::tuple or std::integral_constant. Using C++11 features is only a
 // warning in modern Clang, which are ignored in system headers.
-template<typename> struct __em_asm_sig {};
+template<typename, typename = void> struct __em_asm_sig {};
 template<> struct __em_asm_sig<float> { static const char value = 'd'; };
 template<> struct __em_asm_sig<double> { static const char value = 'd'; };
 template<> struct __em_asm_sig<char> { static const char value = 'i'; };
@@ -113,6 +113,13 @@ template<> struct __em_asm_sig<unsigned long> { static const char value = 'i'; }
 template<> struct __em_asm_sig<bool> { static const char value = 'i'; };
 template<> struct __em_asm_sig<wchar_t> { static const char value = 'i'; };
 template<typename T> struct __em_asm_sig<T*> { static const char value = 'i'; };
+
+// Explicit support for enums, they're passed as int via variadic arguments.
+template<bool> struct __em_asm_if { };
+template<> struct __em_asm_if<true> { typedef void type; };
+template<typename T> struct __em_asm_sig<T, typename __em_asm_if<__is_enum(T)>::type> {
+    static const char value = 'i';
+};
 
 // Instead of std::tuple
 template<typename... Args>

--- a/tests/core/test_em_asm_arguments_side_effects.cpp
+++ b/tests/core/test_em_asm_arguments_side_effects.cpp
@@ -1,0 +1,18 @@
+// Copyright 2019 The Emscripten Authors.  All rights reserved.
+// Emscripten is available under two separate licenses, the MIT license and the
+// University of Illinois/NCSA Open Source License.  Both these licenses can be
+// found in the LICENSE file.
+
+#include <stdio.h>
+#include <emscripten.h>
+
+int main(int argc, char **argv) {
+  int counter_1 = 3, counter_2 = 4;
+  // https://github.com/emscripten-core/emscripten/issues/9030
+  int result = EM_ASM_INT({
+     return $0 * 10 + $1;
+  }, counter_1++, counter_2++);
+  printf("counter_1=3, counter_2=4\n");
+  printf("counter_1++ * 10 + counter_2++=%d\n", result);
+  return 0;
+}

--- a/tests/core/test_em_asm_arguments_side_effects.cpp
+++ b/tests/core/test_em_asm_arguments_side_effects.cpp
@@ -9,10 +9,11 @@
 int main(int argc, char **argv) {
   int counter_1 = 3, counter_2 = 4;
   // https://github.com/emscripten-core/emscripten/issues/9030
+  printf("counter_1=%d, counter_2=%d\n", counter_1, counter_2);
   int result = EM_ASM_INT({
      return $0 * 10 + $1;
   }, counter_1++, counter_2++);
-  printf("counter_1=3, counter_2=4\n");
   printf("counter_1++ * 10 + counter_2++=%d\n", result);
+  printf("counter_1=%d, counter_2=%d\n", counter_1, counter_2);
   return 0;
 }

--- a/tests/core/test_em_asm_arguments_side_effects.out
+++ b/tests/core/test_em_asm_arguments_side_effects.out
@@ -1,0 +1,2 @@
+counter_1=3, counter_2=4
+counter_1++ * 10 + counter_2++=34

--- a/tests/core/test_em_asm_arguments_side_effects.out
+++ b/tests/core/test_em_asm_arguments_side_effects.out
@@ -1,2 +1,3 @@
 counter_1=3, counter_2=4
 counter_1++ * 10 + counter_2++=34
+counter_1=4, counter_2=5

--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -1795,9 +1795,8 @@ int main(int argc, char **argv) {
     self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_parameter_pack')
 
   def test_em_asm_arguments_side_effects(self):
-    self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_arguments_side_effects', force_c=True)
-    self.emcc_args += ['-std=c++11']
     self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_arguments_side_effects')
+    self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_arguments_side_effects', force_c=True)
 
   @parameterized({
     'normal': ([],),

--- a/tests/test_core.py
+++ b/tests/test_core.py
@@ -1794,6 +1794,11 @@ int main(int argc, char **argv) {
     self.emcc_args += ['-std=c++11']
     self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_parameter_pack')
 
+  def test_em_asm_arguments_side_effects(self):
+    self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_arguments_side_effects', force_c=True)
+    self.emcc_args += ['-std=c++11']
+    self.do_run_in_out_file_test('tests', 'core', 'test_em_asm_arguments_side_effects')
+
   @parameterized({
     'normal': ([],),
     'linked': (['-s', 'MAIN_MODULE'],),

